### PR TITLE
Story/ccls 1975 proceedings and costs - proceeding details screens

### PR DESCRIPTION
**Pull request by PhilDigitalJustice**

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+   * Determines if the Validator supports the provided class.
+   *
+   * @param clazz The class to check for support.
+   * @return {@code true} if the class is assignable from
+   *         {@link uk.gov.laa.ccms.caab.bean.proceeding.ProceedingFlowFormData},
+   *         {@code false} otherwise.
+   */
+  @Override
+  public boolean supports(final Class<?> clazz) {
+    return ProceedingFlowFormData.class.isAssignableFrom(clazz);
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java:26]: Should this not be ProceedingFormDataFurtherDetails.class, as that is the target of the validator?

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+   * Validates the further details in the
+   * {@link uk.gov.laa.ccms.caab.bean.proceeding.ProceedingFlowFormData}.
+   *
+   * @param target The object to be validated.
+   * @param errors The Errors object to store validation errors.
+   */
+  @Override
+  public void validate(final Object target, final Errors errors) {
+    final ProceedingFlowFormData formData = (ProceedingFlowFormData) target;
+    final ProceedingFormDataFurtherDetails furtherDetails = formData.getFurtherDetails();
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java:39]: Pass in the ProceedingFormDataFurtherDetails directly, as per ProceedingMatterTypeDetailsValidator

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java [hunk 3]
```diff
@@ -216,6 +216,29 @@ public Mono<CostStructure> getCosts(
...
 
+  /**
+   * Updates the cost structure of an application using the CAAB API.
+   *
+   * @param id The ID of the application.
+   * @param costs The new cost structure for the application.
+   * @param loginId The ID associated with the user login.
+   * @return A Mono Void indicating the completion of the update operation.
+   */
+  public Mono<Void> updateCosts(
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java:227]: I know terminology varies, but maybe better to name it updateCostStructure to match the endpoint

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 4]
```diff
@@ -285,6 +306,76 @@ public Mono<ProceedingDetail> getProceeding(final String proceedingCode) {
...
+   *
+   * @return A Mono containing the ProceedingDetails or an error handler if an error occurs.
+   */
+  public Mono<ProceedingDetails> getProceedings(
+      final String categoryOfLaw,
+      final String matterType,
+      final Boolean isAmendment,
+      final Boolean larScopeFlag,
+      final String applicationType,
+      final boolean isLead) {
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:320]: Is this primitive for a reason?

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 5]
```diff
@@ -285,6 +306,76 @@ public Mono<ProceedingDetail> getProceeding(final String proceedingCode) {
...
+    return ebsApiWebClient
+        .get()
+        .uri(builder -> builder.path("/proceedings")
+            .queryParamIfPresent("category-of-law", Optional.ofNullable(categoryOfLaw))
+            .queryParamIfPresent("matter-type", Optional.ofNullable(matterType))
+            .queryParamIfPresent("amendment-only", Optional.ofNullable(isAmendment))
+            .queryParamIfPresent("lar-scope-flag", Optional.ofNullable(larScopeFlag))
+            .queryParamIfPresent("application-type", Optional.ofNullable(applicationType))
+            .queryParamIfPresent("lead", Optional.of(isLead))
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:329]: If it's primitive, it has to be present?

File: src/main/java/uk/gov/laa/ccms/caab/constants/SessionConstants.java [hunk 6]
```diff
@@ -103,4 +108,20 @@ public class SessionConstants {
 
   public static final String APPLICATION_CLIENT_NAMES = "applicationClientNames";
 
+
+  public static final String APPLICATION_COSTS = "costs";
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/constants/SessionConstants.java:112]: javadocs

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 7]
```diff
@@ -25,63 +61,679 @@
...
-
-          return "application/summary/proceedings-and-costs-section";
-        }).block();
+    final ApplicationDetail application =
+        applicationService.getApplication(applicationId).block();
+
+    applicationService.prepareProceedingSummary(applicationId, application, user);
+    model.addAttribute(APPLICATION, application);
+
+    if (application != null) {
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:111]: If the application was null, would that not be an error condition? To be thrown earlier
Optional.ofNullable(applicationService.getApplication....).block()).orElseThrow(....)

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 8]
```diff
@@ -25,63 +61,679 @@
...
+   *
+   * @param applicationId The id of the application, retrieved from the session.
+   * @param application The application details, retrieved from the session.
+   * @param proceedings The list of proceedings, retrieved from the session.
+   * @param user The UserDetail object representing the user, retrieved from the session.
+   * @param proceedingId The id of the proceeding to retrieve the summary for.
+   * @param model The Model object to add attributes to for the view.
+   * @return The name of the view to be rendered.
+   */
+  @GetMapping("/application/summary/proceedings/{proceeding-id}/summary")
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:170]: We definitely need to revisit these endpoints. Having summary twice... :-)

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 9]
```diff
@@ -25,63 +61,679 @@
...
+            proceedingFlow.getProceedingDetails(),
+            proceedingFlow.getMatterTypeDetails().getMatterType());
+        model.addAttribute(PROCEEDING_FLOW_FORM_DATA, proceedingFlow);
+        return "application/summary/proceedings-further-details";
+      }
+
+      proceedingFlow.setAmended(Boolean.TRUE);
+    } else {
+
+      final ProceedingFlowFormData oldProceedingFlow =
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:491]: Does the proceedingFlow not need to be validated for an edit?

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 10]
```diff
@@ -25,63 +61,679 @@
...
+
+      model.addAttribute(PROCEEDING_FLOW_FORM_DATA, proceedingFlow);
+      model.addAttribute(CURRENT_PROCEEDING, proceeding);
+
+    } else {
+      //Came from previous step in the flow
+      proceedingFlow = (ProceedingFlowFormData) session.getAttribute(PROCEEDING_FLOW_FORM_DATA);
+    }
+
+    if (proceedingFlow.isAmended()) {
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:596]: This section could do with a few comments. I'm struggling to work out what it's doing.

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 11]
```diff
@@ -534,55 +544,204 @@ public ResultsDisplay<LinkedCaseResultRowDisplay> getLinkedCases(final String ap
...
+    //
+    //      // for (Proceeding caseProceeding : myCase.getProceedings()) {
+    //      //   if (caseProceeding.getEbsId().equalsIgnoreCase(proceeding.getEbsId())) {
+    //      //     existingStage = caseProceeding.getStage();
+    //      //   }
+    //      // }
+    //      // see - getAmendmentProceedingStage
+    //    }
+
+    if (scopeLimitations.size() == 1) {
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:673]: Can this block not be removed, and the following bit is only done once if there's only one scopeLimitation?

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 12]
```diff
@@ -1385,4 +1543,69 @@ private String findAwardType(
...
+   * @param user The user performing the operation.
+   */
+  public void prepareProceedingSummary(
+      final String id,
+      final ApplicationDetail application,
+      final UserDetail user) {
+
+    getDefaultCostLimitation(application);
+
+    if (!application.getAmendment()
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:1561]: Now that it's a Boolean, does this need to be Boolean.TRUE.equals(application.getAmendment()) to protect against NullPointerException?

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 13]
```diff
@@ -1385,4 +1543,69 @@ private String findAwardType(
...
+          .setRequestedCostLimitation(application.getCosts().getDefaultCostLimitation());
+    }
+
+    caabApiClient.updateCosts(id, application.getCosts(), user.getLoginId()).block();
+  }
+
+  private BigDecimal getDefaultCostLimitation(final ApplicationDetail application) {
+    BigDecimal defaultCostLimitation = new BigDecimal("0.00");
+
+    final BigDecimal currentDefault = application.getCosts().getDefaultCostLimitation();
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:1573]: Same comment re: getCosts()

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 14]
```diff
@@ -236,6 +244,112 @@ public Mono<CommonLookupDetail> getApplicationStatuses() {
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_MATTER_TYPES);
+  }
+
+  /**
+   * Get a list of proceeding types for proceedings.
+   *
+   * @return CommonLookupDetail containing the common lookup values.
+   */
+  public Mono<ProceedingDetails> getProceedings(
+      final ApplicationDetail application,
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:271]: Could this be changed to just take in the flags and ids that it needs? 
It probably doesn't matter, it's just that so far the LookupService hasn't dealt in Applications/Cases and is kept as standalone as possible.

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 15]
```diff
@@ -236,6 +244,112 @@ public Mono<CommonLookupDetail> getApplicationStatuses() {
...
+   * The level of service types are fetched based on the proceeding code, category of law, and
+   * matter type.
+   *
+   * @param application The application detail object which contains the category of law.
+   * @param proceedingCode The code of the proceeding.
+   * @param matterType The type of the matter.
+   * @return A Mono of LevelOfServiceLookupDetail containing the level of service types.
+   */
+  public Mono<LevelOfServiceLookupDetail> getProceedingLevelOfServiceTypes(
+      final ApplicationDetail application,
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:322]: Same comment here - could we just pass in the categoryOfLawId?

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 16]
```diff
@@ -236,6 +244,112 @@ public Mono<CommonLookupDetail> getApplicationStatuses() {
...
+   *
+   * @return String containing the description of an order type.
+   */
+  public Mono<String> getOrderTypeDescription(final String code) {
+    return ebsApiClient.getCommonValues(COMMON_VALUE_PROCEEDING_ORDER_TYPE)
+        .map(commonLookupDetail -> commonLookupDetail.getContent().stream()
+            .filter(commonLookupValueDetail -> commonLookupValueDetail.getCode().equals(code))
+            .findFirst()
+            .map(CommonLookupValueDetail::getDescription)
+            .orElse(null));
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:351]: Could default to the passed in code, rather than returning null

**Review by porritta — CHANGES_REQUESTED** [review]
Review finally complete

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java (hunk 1 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java:26]: This one requires access to the proceeding details form data to be able to check if the order type is a required feild

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java (hunk 2 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingFurtherDetailsValidator.java:39]: see comment aboce

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingTypeDetailsValidator.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.validation.Errors;
+import uk.gov.laa.ccms.caab.bean.proceeding.ProceedingFormDataMatterTypeDetails;
+import uk.gov.laa.ccms.caab.bean.proceeding.ProceedingFormDataProceedingDetails;
+import uk.gov.laa.ccms.caab.bean.validators.AbstractValidator;
+
+/**
+ * Validator the proceeding details provided by proceeding flow forms.
+ */
+@Component
+public class ProceedingTypeDetailsValidator extends AbstractValidator {
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/proceedings/ProceedingTypeDetailsValidator.java:13]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java (hunk 3 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClient.java:227]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 4 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:320]: it shouldnt be, amended

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 5 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:329]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 18]
```diff
@@ -285,6 +306,76 @@ public Mono<ProceedingDetail> getProceeding(final String proceedingCode) {
...
+   * Retrieves Client involvement types.
+   *
+   * @return A Mono containing the ClientInvolvementTypeLookupDetail or an error handler if an error
+   *         occurs.
+   */
+  public Mono<ClientInvolvementTypeLookupDetail> getClientInvolvementTypes(
+      final String proceedingCode) {
+    return ebsApiWebClient
+        .get()
+        .uri(builder -> builder.path("/lookup/proceeding-client-involvement-types")
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:347]: will raise as an issue separately 

https://github.com/ministryofjustice/laa-ccms-data-api/issues/87

File: src/main/java/uk/gov/laa/ccms/caab/constants/SessionConstants.java (hunk 6 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/constants/SessionConstants.java:112]: added jdoc

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java (hunk 7 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:111]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java (hunk 9 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:491]: Amended the if statement and moved validation before if

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java (hunk 10 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:596]: Added comments

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 11 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:673]: removed if

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 12 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:1561]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 13 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:1573]: data is defaulted

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 14 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:271]: amended

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 15 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:322]: amended

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 16 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:351]: amended

**Review by porritta — APPROVED** [review]
LGTM

**Pull request merged**